src/main/java/loqor/ait/core/engine/block/multi/MultiBlockStructure.java

```diff
@@ -0,0 +1,209 @@
+            return created;
+        }
+
+        // double iterationwow
+        for (StructureTemplate.StructureBlockInfo info : list) {
+            if (info.state().isOf(AITBlocks.GENERIC_SUBSYSTEM)) continue;
+            if (info.state().isAir()) continue;
+
+            BlockPos offset = info.pos().subtract(center);
+            BlockOffset blockOffset = new BlockOffset(new AllowedBlocks(info.state().getBlock()), offset);
+            created.add(blockOffset);
+        }
+
+        AITMod.LOGGER.info("Loaded multiblock structure {} with {} blocks", structure, created.size());
+
+        return created;
+    }
+
+    public record BlockOffset(AllowedBlocks block, BlockPos offset) {
+        public BlockOffset(Block block, int x, int y, int z) {
+            this(new AllowedBlocks(block), new BlockPos(x, y, z));
+        }
+        public BlockOffset(Block block) {
+            this(new AllowedBlocks(block), BlockPos.ORIGIN);
+        }
+
+        public BlockOffset offset(int x, int y, int z) {
+            return new BlockOffset(this.block, this.offset.add(x, y, z));
+        }
+        public BlockOffset offset(BlockPos offset) {
+            return new BlockOffset(this.block, this.offset.add(offset));
+        }
+        public BlockOffset allow(Block... blocks) {
+            this.block.addAll(List.of(blocks));
+            return this;
+        }
+
+        @Override
+        public String toString() {
+            return "BlockOffset{" +
+                    "block=" + block +
+                    ", offset=" + offset +
+                    '}';
+        }
+
+        public static List<BlockOffset> corners(Block block, int x, int y, int z) {
+            return List.of(
+                    new BlockOffset(block, x, y, z),
+                    new BlockOffset(block, -x, y, z),
+                    new BlockOffset(block, x, -y, z),
+                    new BlockOffset(block, x, y, -z),
+                    new BlockOffset(block, -x, -y, z),
+                    new BlockOffset(block, x, -y, -z),
+                    new BlockOffset(block, -x, y, -z),
+                    new BlockOffset(block, -x, -y, -z)
+            );
+        }
+        public static List<BlockOffset> volume(Block block, int width, int height, int depth) {
+            List<BlockOffset> offsets = new ArrayList<>();
+            for (int x = 0; x < width; x++) {
+                for (int y = 0; y < height; y++) {
+                    for (int z = 0; z < depth; z++) {
+                        offsets.add(new BlockOffset(block, x, y, z));
+                    }
+                }
+            }
+            return offsets;
+        }
+        public static List<BlockOffset> square(Block block, Direction dir, int space, @Nullable Block... centre) {
+            if (centre == null) {
+                centre = new Block[]{block};
+            }
+            Block first = centre[0];
```

Comment:
is there... any reason for `centre` to be an array?

Reply:
leftover from old code

Reply:
I see